File: src/Python/killer.py
import sys
import csv
from datetime import datetime

related = [[None]*81 for _ in range(81)]

def get_parms():
	global parms
	parms = {}
	with open('puzzle.csv') as csvfile:
		parm_rows = csv.reader(csvfile, delimiter=',', quotechar='|')
		for row in parm_rows:
			parms[int(row[len(row)-1])] = row

def check_row(aRow,indexOfZero,newVal):
	if indexOfZero in parms:
		parm = parms[indexOfZero]        
		cells = parm[1:len(parm)]
		total = int( parm[0])
		sum = int(newVal)
		for t in cells:
			cell = (int(aRow[int(t)]))
			sum = sum + cell
		if sum != total: return False
	return True        

def recursive_check(a):
	global b
	i = a.find('0')
	if i == -1:
		dateTimeObj = datetime.now()
		timestampStr = dateTimeObj.strftime("%d-%b-%Y (%H:%M:%S.%f)")
		print('End Timestamp   : ', timestampStr)
		print(a)
		sys.exit(0)
		
	excluded_numbers = set()
	for j in range(i):
		if related[i][j] == None:
			if (i//9 == j//9) or (i-j) % 9 == 0 or (i//27 == j//27 and i%9//3 == j%9//3):
				related[i][j] = 1
			else:
				related[i][j] = 0
		if related[i][j] == 1:		
			if a[j] != '0':
				excluded_numbers.add(a[j])

	for m in '123456789':
		if m not in excluded_numbers:
			if check_row(a,i,m):
				b = a[:i]+m+a[i+1:]
				recursive_check(b)
      
if __name__ == '__main__':
	board = "000000000000000000000000000000000000000000000000000000000000000000000000000000000"
	dateTimeObj = datetime.now()
	timestampStr = dateTimeObj.strftime("%d-%b-%Y (%H:%M:%S.%f)")
	print('Start Timestamp : ', timestampStr)
	get_parms()
	recursive_check(board)		

File: src/R/Killer.R
print("Started...")
board <- rep(0, times=81)
rules <- list()

parms <- read.csv('puzzle.csv', header=FALSE)
for (i in 1:nrow(parms)) {
    parm <- parms[i,] + 1
    parm[1] <- parm[1] - 1
    parmArray <- parm[!is.na(parm)]
    
    rules[tail(parmArray, 1)] <- list(parmArray)
}

checkRow <- function(board, cell, val) {
    rule <- rules[[cell]]
    
    if (!is.null(rule)) {
        for (i in 2:length(rule)) {
            val <- val + board[rule[i]]
        }
       
        if (rule[1] != val) {
            return(FALSE)
        }
    }
    return(TRUE)
}

recursiveCheck <- function(board) {
    #print(board, max.levels = NULL, width = 180)
    i <- match(0, board)    
    if (is.na(i)) {
        print(board)
        stop("Finished!")
        
    }

    iz <- i - 1
    
    excludedNumbers <- list()
    for (j in 0:iz) {
        if ((iz%/%9 == j%/%9) | (iz-j) %% 9 == 0 | (iz%/%27 == j%/%27 & iz%%9%/%3 == j%%9%/%3)) {
            if (board[j + 1] != 0) {
                excludedNumbers[length(excludedNumbers) + 1] <- board[j + 1]
            }
        }
    }

    for (m in 1:9) {
        if (is.na(match(m, excludedNumbers))) {
            if (checkRow(board, i, m)) {
                newBoard <- board
                newBoard[i] = m
                recursiveCheck(newBoard)
            }
        }
    }
}

recursiveCheck(board)